File: src/FlashcardParser/SingleLineParser/Command.ts
import { CommandsFactory } from "@/commands/CommandsFactory";
import {
    COMMAND_SEPARATOR,
    INLINE_SUBCOMMAND_START_SYMBOL,
    INLINE_SUBCOMMAND_END_SYMBOL
} from "../Symbols";
import { LineDescriptor, IStudySet } from "../Types/Types";

export function parseCommandLine(
    lineDescriptor: LineDescriptor,
    studySet: IStudySet,
    subCommand?: string,
    inLine: boolean = false
): boolean {
    let { trimmedLine, index } = lineDescriptor;

    if (subCommand) {
        console.log(`[studySet] Start parsing command as sub-command ${subCommand}`);
        trimmedLine = subCommand;
    }

    let endOfCommand = -1;
    let startOfArg = -1;
    let subCommands = [];
    let openCount = 0;
    let openIndex = -1;
    let closeIndex = -1;
    
    for (let i = 0; i < trimmedLine.length; i++) {
        const char = trimmedLine[i];

        if (char == COMMAND_SEPARATOR && openCount === 0) {
            if (endOfCommand === -1) {
                endOfCommand = i;
            }
            startOfArg = i + 1;
            break;
        }

        if (char == INLINE_SUBCOMMAND_START_SYMBOL) {
            if (openCount === 0) {
                openIndex = i;
                if (endOfCommand === -1) {
                    endOfCommand = i;
                }
            }
            openCount++;
        } else if (char == INLINE_SUBCOMMAND_END_SYMBOL) {
            openCount--;
            if (openCount === 0) {
                closeIndex = i;
                const subCommand = trimmedLine.slice(openIndex + 1, closeIndex);
                console.log(`[studySet] Found sub-commands ${subCommand}`);
                subCommands.push(subCommand);
            }
        }
    }


    let cmdName: string;
    let cmdArg: string;
    if (startOfArg === -1) {
        cmdName = trimmedLine;
        cmdArg = null;
        console.log(`[studySet] Found command ${cmdName} with no arg`);
    } else {
        cmdName = trimmedLine.slice(0, endOfCommand).trim();
        cmdArg = trimmedLine.slice(startOfArg).trim();
        console.log(`[studySet] Found command ${cmdName} with arg ${cmdArg}`);
    }

    const command = CommandsFactory.Make(lineDescriptor, studySet, cmdName, cmdArg);

    if (!command) {
        console.error(`[studySet] Unrecognized command at line ${index}: "${trimmedLine}"`);
        return false;
    }

    let receiver = inLine ? lineDescriptor : lineDescriptor.parent;
    receiver.subParts.push({ ...command.toJson(), subParts: [] });

    subCommands.forEach((x) => {
        const ld: LineDescriptor = {
            index: lineDescriptor.index,
            trimmedLine: x,
            originalLine: lineDescriptor.originalLine,
            tabs: lineDescriptor.tabs + 1,
            category: lineDescriptor.category,
            tabbedUnder: [],
            parent: lineDescriptor,
            isComment: lineDescriptor.isComment,
            subParts: []
        };
        parseCommandLine(ld, studySet, x);
    });

    return true;
}

File: service/flashcard_parser/single_line_parser/command.py
from typing import List

from ..symbols import (
    COMMAND_SEPARATOR,
    INLINE_SUBCOMMAND_START_SYMBOL,
    INLINE_SUBCOMMAND_END_SYMBOL,
)
from ..types import LineDescriptor, StudySet
from ..commands.commands_factory import CommandsFactory


def parse_command_line(
    line_descriptor: LineDescriptor,
    study_set: StudySet,
    sub_command: str | None = None,
    in_line: bool = False,
) -> bool:
    trimmed_line = sub_command if sub_command is not None else line_descriptor.trimmed_line

    end_of_command = -1
    start_of_arg = -1
    sub_commands: List[str] = []
    open_count = 0
    open_index = -1
    close_index = -1

    for i, char in enumerate(trimmed_line):
        if char == COMMAND_SEPARATOR and open_count == 0:
            if end_of_command == -1:
                end_of_command = i
            start_of_arg = i + 1
            break
        if char == INLINE_SUBCOMMAND_START_SYMBOL:
            if open_count == 0:
                open_index = i
                if end_of_command == -1:
                    end_of_command = i
            open_count += 1
        elif char == INLINE_SUBCOMMAND_END_SYMBOL:
            open_count -= 1
            if open_count == 0:
                close_index = i
                sub_cmd = trimmed_line[open_index + 1 : close_index]
                print(f"[studySet] Found sub-commands {sub_cmd}")
                sub_commands.append(sub_cmd)

    if start_of_arg == -1:
        cmd_name = trimmed_line.strip()
        cmd_arg = None
        print(f"[studySet] Found command {cmd_name} with no arg")
    else:
        cmd_name = trimmed_line[:end_of_command].strip()
        cmd_arg = trimmed_line[start_of_arg:].strip()
        print(f"[studySet] Found command {cmd_name} with arg {cmd_arg}")

    command = CommandsFactory.make(line_descriptor, study_set, cmd_name, cmd_arg)
    if not command:
        print(f"[studySet] Unrecognized command at line {line_descriptor.index}: '{trimmed_line}'")
        return False

    receiver = line_descriptor if in_line else line_descriptor.parent
    if receiver is None:
        print("[studySet] Command has no parent to attach to")
        return False
    receiver.sub_parts.append({**command.to_json(), "subParts": [], "sourceIndex": line_descriptor.index})

    for sub in sub_commands:
        ld = LineDescriptor(
            index=line_descriptor.index,
            trimmed_line=sub,
            original_line=line_descriptor.original_line,
            tabs=line_descriptor.tabs + 1,
            category=line_descriptor.category,
            tabbed_under=[],
            parent=line_descriptor,
            is_comment=line_descriptor.is_comment,
            sub_parts=[],
        )
        parse_command_line(ld, study_set, sub)
    return True
